File: frontend/src/services/genai.js
/**
 * Enterprise Architecture Solution - GenAI Service
 * 
 * This module provides functions for interacting with the GenAI API endpoints.
 */

import axios from 'axios';

const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000';

/**
 * Create an API client with proper headers
 * @returns {Object} Axios instance
 */
const createApiClient = () => {
  const token = localStorage.getItem('auth_token');
  
  return axios.create({
    baseURL: `${API_URL}/api/genai`,
    headers: {
      'Content-Type': 'application/json',
      'Authorization': token ? `Bearer ${token}` : '',
    },
  });
};

/**
 * Generate documentation for an EA artifact
 * @param {Object} params - Documentation parameters
 * @param {string} params.contentType - Type of content (element, model, view, policy)
 * @param {string} params.contentId - UUID of the content
 * @param {string} params.format - Output format (markdown, html, docx)
 * @param {boolean} params.includeDiagrams - Whether to include diagrams
 * @param {boolean} params.includeRelationships - Whether to include relationships
 * @param {string} params.style - Style of documentation (technical, business, executive)
 * @returns {Promise<Object>} Generated documentation and metadata
 */
export const generateDocumentation = async ({
  contentType, 
  contentId, 
  format = 'markdown',
  includeDiagrams = true,
  includeRelationships = true,
  style = 'technical'
}) => {
  try {
    const apiClient = createApiClient();
    const response = await apiClient.post('/documentation', {
      content_type: contentType,
      content_id: contentId,
      format,
      include_diagrams: includeDiagrams,
      include_relationships: includeRelationships,
      style
    });
    
    return response.data;
  } catch (error) {
    console.error('Error generating documentation:', error);
    throw error;
  }
};

/**
 * Analyze the impact of a proposed change to an architecture element
 * @param {Object} params - Impact analysis parameters
 * @param {string} params.elementId - UUID of the element being changed
 * @param {string} params.changeDescription - Description of the proposed change
 * @param {string} params.changeType - Type of change (modify, replace, remove)
 * @param {number} params.analysisDepth - Depth of impact analysis (1=direct, 2=indirect, 3=comprehensive)
 * @returns {Promise<Object>} Impact analysis results
 */
export const analyzeImpact = async ({
  elementId,
  changeDescription,
  changeType,
  analysisDepth = 2
}) => {
  try {
    const apiClient = createApiClient();
    const response = await apiClient.post('/impact-analysis', {
      element_id: elementId,
      change_description: changeDescription,
      change_type: changeType,
      analysis_depth: analysisDepth
    });
    
    return response.data;
  } catch (error) {
    console.error('Error analyzing impact:', error);
    throw error;
  }
};

/**
 * Recognize patterns in architecture elements
 * @param {Object} params - Pattern recognition parameters
 * @param {string} params.modelId - UUID of the model to analyze
 * @param {Array<string>} params.elementIds - Optional list of specific element IDs to analyze
 * @param {string} params.domainFilter - Optional domain to filter by
 * @param {Array<string>} params.patternTypes - Optional list of pattern types to look for
 * @returns {Promise<Object>} Recognized patterns
 */
export const recognizePatterns = async ({
  modelId,
  elementIds,
  domainFilter,
  patternTypes
}) => {
  try {
    const apiClient = createApiClient();
    const response = await apiClient.post('/pattern-recognition', {
      model_id: modelId,
      element_ids: elementIds,
      domain_filter: domainFilter,
      pattern_types: patternTypes
    });
    
    return response.data;
  } catch (error) {
    console.error('Error recognizing patterns:', error);
    throw error;
  }
};

/**
 * Run the EA Assistant with the given conversation
 * @param {Array<Object>} messages - List of conversation messages with role and content
 * @returns {Promise<Object>} Assistant's response
 */
export const runAssistant = async (messages) => {
  try {
    const apiClient = createApiClient();
    const response = await apiClient.post('/assistant', {
      messages: messages.map(msg => ({
        role: msg.role,
        content: msg.content
      }))
    });
    
    return response.data;
  } catch (error) {
    console.error('Error running assistant:', error);
    throw error;
  }
};

/**
 * Helper function to export GenAI services
 */
export default {
  generateDocumentation,
  analyzeImpact,
  recognizePatterns,
  runAssistant
};

File: backend/routers/genai.py
"""
Enterprise Architecture Solution - GenAI API Router

This module defines the FastAPI router for GenAI features.
"""

import logging
from typing import Dict, List, Any, Optional
from fastapi import APIRouter, Depends, HTTPException, Security, status
from pydantic import BaseModel

# Import the GenAI service
from ..genai import GenAIService

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Create router
router = APIRouter()

# Pydantic models for request validation
class DocumentationRequest(BaseModel):
    content_type: str
    content_id: str
    format: str = "markdown"
    include_diagrams: bool = True
    include_relationships: bool = True
    style: str = "technical"

class ImpactAnalysisRequest(BaseModel):
    element_id: str
    change_description: str
    change_type: str
    analysis_depth: int = 2

class PatternRecognitionRequest(BaseModel):
    model_id: str
    element_ids: Optional[List[str]] = None
    domain_filter: Optional[str] = None
    pattern_types: Optional[List[str]] = None

class AssistantMessage(BaseModel):
    role: str
    content: str

class AssistantRequest(BaseModel):
    messages: List[AssistantMessage]

# Dependency to get GenAI service
def get_genai_service():
    # This would normally be initialized with a Supabase client
    # from a dependency injection system
    from supabase import create_client
    import os
    
    supabase_url = os.getenv("SUPABASE_URL")
    supabase_key = os.getenv("SUPABASE_KEY")
    
    if not supabase_url or not supabase_key:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Supabase configuration is missing"
        )
    
    supabase = create_client(supabase_url, supabase_key)
    return GenAIService(supabase)

# Route for documentation generation
@router.post("/documentation", tags=["documentation"])
async def generate_documentation(
    request: DocumentationRequest,
    genai_service: GenAIService = Depends(get_genai_service)
):
    """
    Generate documentation for an EA artifact.
    
    Parameters:
    - content_type: Type of content (element, model, view, policy)
    - content_id: UUID of the content
    - format: Output format (markdown, html, docx)
    - include_diagrams: Whether to include diagrams
    - include_relationships: Whether to include relationships
    - style: Style of documentation (technical, business, executive)
    
    Returns:
    - Generated documentation and metadata
    """
    try:
        result = genai_service.generate_documentation(
            request.content_type,
            request.content_id,
            request.format,
            request.include_diagrams,
            request.include_relationships,
            request.style
        )
        
        if not result["success"]:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=result.get("error", "Failed to generate documentation")
            )
            
        return result
    except Exception as e:
        logger.error(f"Error generating documentation: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )

# Route for impact analysis
@router.post("/impact-analysis", tags=["impact"])
async def analyze_impact(
    request: ImpactAnalysisRequest,
    genai_service: GenAIService = Depends(get_genai_service)
):
    """
    Analyze the impact of a proposed change to an architecture element.
    
    Parameters:
    - element_id: UUID of the element being changed
    - change_description: Description of the proposed change
    - change_type: Type of change (modify, replace, remove)
    - analysis_depth: Depth of impact analysis (1=direct, 2=indirect, 3=comprehensive)
    
    Returns:
    - Impact analysis results
    """
    try:
        result = genai_service.analyze_impact(
            request.element_id,
            request.change_description,
            request.change_type,
            request.analysis_depth
        )
        
        if not result["success"]:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=result.get("error", "Failed to analyze impact")
            )
            
        return result
    except Exception as e:
        logger.error(f"Error analyzing impact: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )

# Route for pattern recognition
@router.post("/pattern-recognition", tags=["patterns"])
async def recognize_patterns(
    request: PatternRecognitionRequest,
    genai_service: GenAIService = Depends(get_genai_service)
):
    """
    Recognize patterns in architecture elements.
    
    Parameters:
    - model_id: UUID of the model to analyze
    - element_ids: Optional list of specific element IDs to analyze
    - domain_filter: Optional domain to filter by
    - pattern_types: Optional list of pattern types to look for
    
    Returns:
    - Recognized patterns
    """
    try:
        result = genai_service.recognize_patterns(
            request.model_id,
            request.element_ids,
            request.domain_filter,
            request.pattern_types
        )
        
        if not result["success"]:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=result.get("error", "Failed to recognize patterns")
            )
            
        return result
    except Exception as e:
        logger.error(f"Error recognizing patterns: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )

# Route for EA Assistant
@router.post("/assistant", tags=["assistant"])
async def run_assistant(
    request: AssistantRequest,
    genai_service: GenAIService = Depends(get_genai_service)
):
    """
    Run the EA Assistant with the given conversation.
    
    Parameters:
    - messages: List of conversation messages with role and content
    
    Returns:
    - Assistant's response
    """
    try:
        # Convert Pydantic models to dicts
        messages = [{"role": msg.role, "content": msg.content} for msg in request.messages]
        
        result = genai_service.run_assistant(messages)
        
        if not result["success"]:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=result.get("error", "Failed to run assistant")
            )
            
        return result
    except Exception as e:
        logger.error(f"Error running assistant: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )
